File: 93-Restore-IP-Addresses/solution.java
public class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> result = new ArrayList<String>();
        if(s.length() < 4 || s.length() > 12) return result;
        dfs(s, result, "", 0, 0);
        return result;
    }
    
    void dfs(String s, List<String> result, String temp, int index, int k) {
        if(index == s.length()) {
            if(k == 4) result.add(temp);
            return;
        }
        for(int i = 1; i <= 3 && index + i <= s.length(); i++) {
            String substr = s.substring(index, index + i);
            if(isValid(substr)) {
                if(temp.length() == 0) dfs(s, result, temp + substr, index + i, k + 1);
                else dfs(s, result, temp + "." + substr, index + i, k + 1);
            }
        }
    }
    
    boolean isValid(String s) {
        if(s == null || s.length() == 0) return false;
        if(s.charAt(0) == '0') return s.equals("0");
        return Integer.valueOf(s) > 0 && Integer.valueOf(s) <= 255;
    }
}

File: 93-Restore-IP-Addresses/solution.py
class Solution(object):
    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        if len(s) < 4 or len(s) > 12:
            return []
        result = []
        self.dfs(s, result, "", 0)
        return result
        
    def dfs(self, s, result, path, count):
        if count == 3:
            if self.isValid(s):
                result.append(path + s)
            return
        i = 1
        while i <= 3 and i < len(s):
            cur = s[: i]
            if self.isValid(cur):
                self.dfs(s[i:], result, path + cur + ".", count + 1)
            i += 1
    
    def isValid(self, cur):
        if cur[0] == '0':
            return cur == "0"
        return 0 < int(cur) <= 255
    